### PR TITLE
[DependencyInjection] Add support an integer return for default_index_method

### PR DESCRIPTION
**Pull request by maranqz**
| Q             | A
| ------------- | ---
| Branch?       | 5.x for features
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | #40319 
| License       | MIT
| Doc PR        | TODO

### CODE REVIEW

**Comment by maranqz** [comment]
Should I put notice about this feature in [docs](https://github.com/symfony/symfony-docs/blame/5.x/service_container/service_subscribers_locators.rst#L558) ?

File: src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php [hunk 1]
```diff
@@ -137,13 +137,7 @@ public static function getDefaultIndex(ContainerBuilder $container, string $serv
...
         }
 
-        $defaultIndex = $rm->invoke(null);
-
-        if (!\is_string($defaultIndex)) {
-            throw new InvalidArgumentException(implode(sprintf('return a string (got "%s")', get_debug_type($defaultIndex)), $message));
-        }
-
-        return $defaultIndex;
+        return (string) $rm->invoke(null);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php]: I would not unconditionally cast to string here. that could have weird effects with other things than integers that you want to support here.

Comment (maranqz) [reply src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php]:
```php
if (\is_int($defaultIndex)){  // or \is_numeric?
    $defaultIndex = (string) $defaultIndex;
}
if (!\is_string($defaultIndex)) {
....
```
Is that better?

File: src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php [hunk 2]
```diff
@@ -139,6 +139,10 @@ public static function getDefaultIndex(ContainerBuilder $container, string $serv
 
         $defaultIndex = $rm->invoke(null);
 
+        if (\is_numeric($defaultIndex)) {
```
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php]:
```suggestion
        if (\is_int($defaultIndex)) {
```

I don't think, floats play well here, so let's explicitly check for integers.

File: src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php [hunk 3]
```diff
@@ -139,6 +139,10 @@ public static function getDefaultIndex(ContainerBuilder $container, string $serv
 
         $defaultIndex = $rm->invoke(null);
 
+        if (is_numeric($defaultIndex)) {
```
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/Compiler/PriorityTaggedServiceTrait.php]:
```suggestion
        if (\is_int($defaultIndex)) {
```

I don't think we want to allow floats as well, so please let's explicitly check for integers.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 4]
```diff
@@ -10,6 +10,7 @@ CHANGELOG
  * Add autoconfigurable attributes
  * Add support for per-env configuration in loaders
  * Add `ContainerBuilder::willBeAvailable()` to help with conditional configuration
+ * Add support an integer return for default_index_method
```
Comment (derrabus) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]:
```suggestion
 * Add support an integer return value for default_index_method
```

**Comment by maranqz** [comment]
I do not understand how I should fix 7.4 tests.
Could you help me with?

image

**Review by Nyholm — APPROVED** [review]
Awesome!

Great feature, great implementation, great that you added both test and change log. 
I like this. 👍 

The test failures is not related to your PR so you can ignore them. Good thing that you checked. 

I dont think this feature needs to be documented. It is almost as I think it should be a bugfix...

Let's see what some other maintainer think about adding docs or not.

**Review by derrabus — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @maranqz.

**Pull request merged**